### ISSUE TITLE
ValueError: dependence_plot error for categorical variables with matplotlib >= 3.5.0

### ISSUE THREAD

**Issue by larpal** [shap/shap#2273]
**Bug Summary**
The function `dependence_plot` throws a value error for interactions with categorical variables for matplotlib version 3.5.0. 
The bug is caused since `matplotlib.pyplot.scatter` prohibits the use of the parameters `norm` and `vmin/vmax` at the same time which the `dependence_plot` does for categorical variables.

**SHAP Version**
0.40.0

**Matplotlib Version**
3.5.0

**Python Version**
Python 3.7.10

**Code for Reproduction**

```
import shap
import matplotlib.pyplot as plt
import numpy as np

n_points = 100

# create two float and one binary feature
X = np.concatenate([np.random.rand(n_points,2), np.random.randn(n_points,1)>0], axis=1)
shap_values = np.random.randn(*X.shape)

# features 0 and 1 are both floats
print("numerical / numerical interaction:")
shap.dependence_plot(0,shap_values,X, interaction_index=1)

# feature 2 is binary
print("numerical / binary interaction:")
shap.dependence_plot(0,shap_values,X, interaction_index=2)
```
**Outcome**
numerical / numerical interaction:
![image](https://user-images.githubusercontent.com/2567456/142423331-3b517800-2c23-4d52-8af8-bf8a10e4ec81.png)
numerical / binary interaction:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-44-ec650db72f40> in <module>
      8 shap.dependence_plot(0,shap_values,X, interaction_index=1)
      9 print("numerical / binary interaction")
---> 10 shap.dependence_plot(0,shap_values,X, interaction_index=2)

/opt/conda/lib/python3.7/site-packages/shap/plots/dependence.py in dependence_plot(ind, shap_values, features, feature_names, display_features, interaction_index, color, axis_color, cmap, dot_size, x_jitter, alpha, title, xmin, xmax, ax, show)

/opt/conda/lib/python3.7/site-packages/matplotlib/__init__.py in inner(ax, data, *args, **kwargs)
   1410     def inner(ax, *args, data=None, **kwargs):
   1411         if data is None:
-> 1412             return func(ax, *map(sanitize_sequence, args), **kwargs)
   1413 
   1414         bound = new_sig.bind(ax, *args, **kwargs)

/opt/conda/lib/python3.7/site-packages/matplotlib/axes/_axes.py in scatter(self, x, y, s, c, marker, cmap, norm, vmin, vmax, alpha, linewidths, edgecolors, plotnonfinite, **kwargs)
   4426             collection.set_cmap(cmap)
   4427             collection.set_norm(norm)
-> 4428             collection._scale_norm(norm, vmin, vmax)
   4429 
   4430         # Classic mode only:

/opt/conda/lib/python3.7/site-packages/matplotlib/cm.py in _scale_norm(self, norm, vmin, vmax)
    379             if norm is not None:
    380                 raise ValueError(
--> 381                     "Passing parameters norm and vmin/vmax simultaneously is "
    382                     "not supported. Please pass vmin/vmax directly to the "
    383                     "norm when creating it.")

ValueError: Passing parameters norm and vmin/vmax simultaneously is not supported. Please pass vmin/vmax directly to the norm when creating it.
```

**Comment by giorgiabergamo** [comment]
+1, any updates on this ?

**Comment by ghost** [comment]
+1, same issue.

**Comment by mariusgarbowski** [comment]
+1
this has something to do with the parameter _interaction_index_ which is set to "auto" by default. If interaction_index is set to some feature name it sometimes works (not always).
Edit: It seems to work, if the feature values are "natural" floating values. Features, which had been transformed to floating dtypes before training does not seem to work.

**Comment by tatesnow** [comment]
+1, same issue.

**Comment by jklaise** [comment]
The change in `matplotlib` causing the failure is described [here](https://matplotlib.org/3.5.0/api/prev_api_changes/api_changes_3.5.0.html#stricter-validation-of-function-parameters) (5th bullet point).

**Comment by OctavioPJ** [comment]
downgrading matplotlib to 3.2.2 worked for me as a temporary solution

**Comment by GuidoBorrelli** [comment]
Same, thanks for the advice

**Comment by felix515** [comment]
+1, same issue, but downgrading matplotlib to 3.2.2 also worked for me.

**Comment by GuidoBorrelli** [comment]
For me 3.4.0 already solved the issue

**Comment by IamJYC** [comment]
+1 I have the same issue, any updates for now?

**Comment by hellenesherman** [comment]
+1 Same issue for me with 3.5.1. Matplotlib

**Comment by detrin** [comment]
+1
I am having the same issue. Everything above 3.3.0 fails. Check the following release https://matplotlib.org/3.3.0/api/api_changes.html

Parameters norm and vmin/vmax should not be used simultaneously ([link](https://matplotlib.org/3.3.0/api/api_changes.html#parameters-norm-and-vmin-vmax-should-not-be-used-simultaneously))

Passing parameters norm and vmin/vmax simultaneously to functions using colormapping such as scatter() and imshow() is deprecated. Inestead of norm=LogNorm(), vmin=min_val, vmax=max_val pass norm=LogNorm(min_val, max_val). vmin and vmax should only be used without setting norm.

I would like to see an update on this

**Comment by matteodelucchi** [comment]
+1 for having the same issue. Downgrading helped temporarily but results in a warning.

For some reason, the labels for the _interaction_index_ are not always displayed (this is not reproducible...). Might be related to @mariusgarbowski's [comment](https://github.com/slundberg/shap/issues/2273#issuecomment-985883669).

Without downgrading `matplotlib`, the result from [`shap.plots.scatter()`](https://github.com/slundberg/shap/blob/9411b68e8057a6c6f3621765b89b24d82bee13d4/shap/plots/_scatter.py#L19) is very similar to the [`shap.dependence_plot()`](https://github.com/slundberg/shap/blob/9411b68e8057a6c6f3621765b89b24d82bee13d4/shap/plots/_scatter.py#L434) ([dependence_legacy -> dependence_plot](https://github.com/slundberg/shap/blob/79354776372e6add7f495fbaf4c660dc4f771de6/shap/__init__.py#L40)) but has no problems with `matplotlib` 3.5.1. Does anyone know how they exactly differ?

**Comment by dex314** [comment]
+1 for the same issue. 
I made a workaround below that sort of imitates the dependence plots...

````
shap_val_df = pd.DataFrame(shap_values[1], index=X.index, columns=X.columns)
shap_val_df['target'] = y

## colormap to kinda match
from matplotlib.colors import LinearSegmentedColormap

colors = [(0, 0.5, 1), (1, 1, 1), (1, 0.2, 0.2)] #Bluish->White->Salmony
cmap_name = 'BlWhSm'
cm = LinearSegmentedColormap.from_list(cmap_name, colors, N=500)
plt.register_cmap(cmap=cm)

for ci in shap_val_df.columns:
    fig, ax=plt.subplots(1,1,figsize=(12,6))
    scatter = ax.scatter(x=X[ci], 
                                   y=shap_val_df[ci], 
                                   c=shap_val_df.target, 
                                   cmap='BlWhSm', alpha=0.6, marker='o', s=5);
    ax.axhline(0, c='k', alpha=0.3)
    ax.axvline(0, c='k', alpha=0.3)
     
    ax.legend(*scatter.legend_elements())
    ax.set_ylabel(f"SHAP VALUE - {ci}", fontsize=12)
    ax.set_xlabel(f"{ci}", fontsize=12);
```

**Comment by AH012** [comment]
Hello, +1 same problem for me

**Comment by poedator** [comment]
It looks like this issue was fixed [by this commit](https://github.com/slundberg/shap/commit/8065cf7f9239574eed1c7d0ab3b40d538d16608d) on Mar 22, 2022. It is not yet part of release (as of 0.40.0). If necessary - just repeat this edit in your local installation ( remove `vmin=clow, vmax=chigh,` in shap/plots/_scatter.py).

**Comment by GregSilverman** [comment]
Same issue. 

Only workable solution was downgrading to matplotlib==3.2.1, which in and of itself was a dependency hell!

**Comment by ChenZhanfei** [comment]
If adding `norm = None` , it will temporarily work.

**Comment by carl-offerfit** [comment]
I got the error with SHAP 0.40.0 but it was fixed by upgrading to 0.41.0. Using matplotlib 3.5.2

**Comment by wxl112** [comment]
> I got the error with SHAP 0.40.0 but it was fixed by upgrading to 0.41.0. Using matplotlib 3.5.2

thanks for your share! I make it by taking your advice!

**Comment by connortann** [comment]
Closing as I believe this is issue was fixed in release `0.41.0`. I cannot reproduce the error on the `master` branch.
